### DISCUSSION TITLE
External packages depending on internal packages

### DISCUSSION THREAD

**Discussion by yamcodes** [vercel/turborepo#7686] [Help]
### Summary

Is it possible to release packages that depend on internal packages in their `dependencies`?

### Additional information

Imagine the following structure:

```
- packages/
  - a-public/
  - b-internal/
```

Let's say we want to publish `a-public`, a package containing code from `b-internal` as it is installed in `a-public`'s `dependencies`:
```
  "dependencies": {
    "@repo/b-internal": "*"
  }
```

Is this a supported flow? We tried achieving this and while the build is not failing, the usage fails since the imports `import { Thing } from "@repo/b-internal";` are failing for referencing a non-existent import. 

Ideally, there would be a flow that outputs `@repo/b-internal` to the `node_modules` of the `@repo/a-external` package.

What are we missing?

**Comment by mehulkar** [comment]
> Is this a supported flow?

To set expectations here, this has unrelated to turborepo. Turborepo does not manage your dependencies or publishing your packages. That said, this _is_ a tricky problem with Monorepos and @anthonyshew is working on guides that will help set some blueprints! 

> the usage fails

I'm assuming you mean usage of the public (`@repo/a-public`) dependency _outside_ of the monorepo (or at least through the npm registry, even if it's in another package in your monorepo).

> Ideally, there would be a flow that outputs `@repo/b-internal` to the `node_modules` of the `@repo/a-external` package.

This seems right to me also! I think the easiest way to achieve this is by bundling`@repo/a-public` before publishing, so that the packed artifact that is published includes its dependencies. You can do this with `esbuild` or `rollup` or a number of other tools in a build step.

**Reply by anthonyshew** [reply]
Just as an example, `tsup` has a `noExternals` flag that should help here. As Mehul noted, other bundlers will have something similar.
